File: natural_language_replace.py
import re
from pathlib import Path

def replace_natural_language(file_path, query, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    edit_map = {
        "replace _init_exchange method": (
            r"def _init_exchange\(self\)(.*?)(?=\n\s*def|\n\s*class|\Z)",
            "def _init_exchange(self):\n    self.exchange = BybitV5Plugin(self.cfg)"
        )
    }
    old_pattern, new_text = edit_map.get(query.lower(), (query, query))
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    pattern = re.compile(old_pattern, re.DOTALL)
    if not pattern.search(content):
        raise ValueError(f"Pattern for '{query}' not found in {file_path}")
    new_content = pattern.sub(new_text, content)
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(new_content)
    return f"Applied natural language replacement '{query}' in {file_path}"

File: ts-tools/natural_language_replace.ts


import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

export function replaceNaturalLanguage(filePath: string, query: string, config: any): Promise<string> {
    return new Promise((resolve, reject) => {
        checkFilePermission(filePath, "write", config)
            .then(() => {
                const resolvedPath = path.resolve(filePath);
                const editMap: { [key: string]: [string, string] } = {
                    "replace _init_exchange method": [
                        `def _init_exchange\\(self\\)(.*?)(?=\\n\\s*def|\\n\\s*class|$)`,
                        "def _init_exchange(self):\n    self.exchange = BybitV5Plugin(self.cfg)"
                    ]
                };
                const [oldPattern, newText] = editMap[query.toLowerCase()] || [query, query];
                fs.readFile(resolvedPath, 'utf-8', (err, data) => {
                    if (err) {
                        reject(err);
                    } else {
                        const regex = new RegExp(oldPattern, 's');
                        if (!regex.test(data)) {
                            reject(new Error(`Pattern for '${query}' not found in ${filePath}`));
                        } else {
                            const newContent = data.replace(regex, newText);
                            fs.writeFile(resolvedPath, newContent, 'utf-8', (err) => {
                                if (err) {
                                    reject(err);
                                } else {
                                    resolve(`Applied natural language replacement '${query}' in ${filePath}`);
                                }
                            });
                        }
                    }
                });
            })
            .catch(reject);
    });
}

